contracts/near/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+//#![deny(warnings)]
+
+use {
+    error::Error,
+    near_sdk::{
+        borsh::{
+            self,
+            BorshDeserialize,
+            BorshSerialize,
+        },
+        collections::{
+            UnorderedMap,
+            UnorderedSet,
+        },
+        env,
+        is_promise_success,
+        log,
+        near_bindgen,
+        AccountId,
+        BorshStorageKey,
+        Gas,
+        PanicOnDefault,
+        Promise,
+    },
+    p2w_sdk::{
+        BatchPriceAttestation,
+        Identifier,
+        PriceStatus,
+    },
+    state::{
+        PriceFeed,
+        Source,
+    },
+    std::io::Cursor,
+};
+
+pub mod error;
+pub mod ext;
+pub mod state;
+
+const GAS_FOR_VERIFY_VAA: Gas = Gas(10_000_000_000_000);
+const STALENESS_THRESHOLD: u64 = 60;
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum StorageKeys {
+    Source,
+    Prices,
+}
+
+/// The `State` contains all persisted state for the contract. This includes runtime configuration.
+///
+/// There is no valid Default state for this contract, so we derive PanicOnDefault to force
+/// deployment using one of the #[init] functions in the impl below.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct State {
+    /// The set of `Source`s from which Pyth attestations are allowed to be relayed from.
+    pub source: UnorderedSet<Source>,
+
+    /// A Mapping from PriceFeed ID to Price Info.
+    pub prices: UnorderedMap<Identifier, PriceFeed>,
+
+    /// The `AccountId` of the Wormhole contract used to verify VAA's.
+    pub wormhole: AccountId,
+
+    /// A hash of the current contract code.
+    ///
+    /// If this `hash` does not match the current contract code, then it indicates that a pending
+    /// upgrade has been allowed, calling `upgrade` with code that matches this hash will cause the
+    /// contract to upgrade itself.
+    pub codehash: [u8; 32],
+}
+
+/// The `Action` acts as an enum of possible `Action`'s this contract can receive via VAA.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub enum Action {
+    AddSource(Source),
+    DelSource(Source),
+    BatchAttest(Vec<u8>),
+    UpgradeContract([u8; 32]),
+}
+
+#[near_bindgen]
+impl State {
+    #[init]
+    #[allow(clippy::new_without_default)]
+    pub fn new(wormhole: AccountId, codehash: [u8; 32]) -> Self {
```

Comment:
I'm not getting how sources could be set when there is no initial source :think: can you explain more?